File: apps/android/app/src/main/java/com/swcsoftware/valuelens/ui/screens/PresentationPickers.kt
package com.swcsoftware.valuelens.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.requiredSize
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.wrapContentSize
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.TransformOrigin
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.swcsoftware.valuelens.core.Grading
import com.swcsoftware.valuelens.core.Lens
import com.swcsoftware.valuelens.core.LensInfo
import com.swcsoftware.valuelens.data.SampleValuationRepository
import com.swcsoftware.valuelens.domain.ValuationModel
import com.swcsoftware.valuelens.domain.ValuationReport
import com.swcsoftware.valuelens.ui.LayoutStyle
import com.swcsoftware.valuelens.ui.layouts.ClassicLayout
import com.swcsoftware.valuelens.ui.layouts.LayoutContext
import com.swcsoftware.valuelens.ui.layouts.ReportCardLayout
import com.swcsoftware.valuelens.ui.theme.VL
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Value or Growth, described in the core's own words (Grading.lenses()), so onboarding and Settings
 * cannot describe the lens differently. Mirrors apps/ios/.../Settings/PresentationPickers.swift.
 */
@Composable
fun LensPicker(selected: Lens, onSelect: (Lens) -> Unit, cards: Boolean = false) {
    val lenses = remember { Grading.lenses() }
    if (cards) {
        Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
            lenses.forEach { info -> LensCard(info, info.key == selected.name) { onSelect(Lens.from(info.key)) } }
        }
    } else {
        Column {
            SingleChoiceSegmentedButtonRow(Modifier.fillMaxWidth()) {
                lenses.forEachIndexed { i, info ->
                    SegmentedButton(
                        selected = info.key == selected.name,
                        onClick = { onSelect(Lens.from(info.key)) },
                        shape = SegmentedButtonDefaults.itemShape(i, lenses.size),
                    ) { Text(info.name) }
                }
            }
            lenses.firstOrNull { it.key == selected.name }?.let {
                Text(it.blurb, style = MaterialTheme.typography.bodySmall, color = VL.textSecondary,
                     modifier = Modifier.padding(top = 8.dp))
            }
        }
    }
}

@Composable
private fun LensCard(info: LensInfo, chosen: Boolean, onClick: () -> Unit) {
    Row(
        Modifier.fillMaxWidth()
            .background(VL.surface, RoundedCornerShape(16.dp))
            .border(if (chosen) 2.dp else 1.dp, if (chosen) VL.accent else VL.border, RoundedCornerShape(16.dp))
            .clickable(onClick = onClick)
            .padding(16.dp)
            .semantics { selected = chosen; contentDescription = "${info.name}. ${info.blurb}" },
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f)) {
            Text(info.name, fontSize = 17.sp, fontWeight = FontWeight.SemiBold, color = VL.textPrimary)
            Text(info.blurb, fontSize = 14.sp, color = VL.textSecondary, modifier = Modifier.padding(top = 6.dp))
        }
        Text(if (chosen) "✓" else "›", fontSize = 20.sp, color = if (chosen) VL.accent else VL.textTertiary,
             modifier = Modifier.padding(start = 12.dp))
    }
}

/**
 * The two layouts, each drawn by the layout itself on the bundled Apple sample and scaled down — so
 * what you see here is what you will get, in the current theme and accent.
 */
@Composable
fun LayoutPicker(selected: LayoutStyle, lens: Lens, onSelect: (LayoutStyle) -> Unit) {
    val context = LocalContext.current
    val sample by produceState<ValuationReport?>(null) {
        value = withContext(Dispatchers.IO) { runCatching { SampleValuationRepository(context).valuation("AAPL") }.getOrNull() }
    }
    Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(14.dp)) {
        LayoutStyle.entries.forEach { style ->
            val chosen = style == selected
            Column(
                Modifier.weight(1f).clickable { onSelect(style) }
                    .semantics(mergeDescendants = true) { this.selected = chosen; contentDescription = "${style.displayName} layout. ${style.blurb}" },
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Thumbnail(style, sample, lens, chosen)
                Text((if (chosen) "✓ " else "") + style.displayName, fontSize = 14.sp, fontWeight = FontWeight.SemiBold,
                     color = if (chosen) VL.accent else VL.textPrimary, modifier = Modifier.padding(top = 8.dp))
                Text(style.blurb, fontSize = 11.sp, color = VL.textSecondary, textAlign = TextAlign.Center)
            }
        }
    }
}

private const val PHONE_WIDTH = 393
private const val CROP_HEIGHT = 640
private const val SCALE = 0.34f

@Composable
private fun Thumbnail(style: LayoutStyle, sample: ValuationReport?, lens: Lens, chosen: Boolean) {
    Box(
        Modifier.size((PHONE_WIDTH * SCALE).dp, (CROP_HEIGHT * SCALE).dp)
            .clip(RoundedCornerShape(10.dp))
            .background(VL.background)
            .border(if (chosen) 2.5.dp else 1.dp, if (chosen) VL.accent else VL.border, RoundedCornerShape(10.dp)),
        contentAlignment = Alignment.Center,
    ) {
        if (sample == null) {
            CircularProgressIndicator(Modifier.size(20.dp))
        } else {
            val ctx = remember(sample, lens) {
                LayoutContext(
                    report = sample, result = sample.modelA, expert = false, expertModeAlwaysOn = false,
                    model = ValuationModel.A, onModel = {}, onEditPrice = {}, onSetShowMath = {},
                    onReportIssue = {}, reportCard = Grading.reportCard(sample, lens),
                )
            }
            // Laid out at phone width, then scaled — the layout never knows it is a thumbnail.
            Box(
                Modifier.wrapContentSize(Alignment.TopStart, unbounded = true)
                    .requiredSize(PHONE_WIDTH.dp, CROP_HEIGHT.dp)
                    .graphicsLayer { scaleX = SCALE; scaleY = SCALE; transformOrigin = TransformOrigin(0f, 0f) },
            ) {
                when (style) {
                    LayoutStyle.CLASSIC -> ClassicLayout(ctx)
                    LayoutStyle.REPORT_CARD -> ReportCardLayout(ctx)
                }
            }
            // Catches every touch so the miniature cannot be scrolled or tapped into.
            Box(Modifier.matchParentSize().clickable(enabled = false) {})
        }
    }
}

File: apps/ios/ValueLens/Features/Settings/PresentationPickers.swift
import SwiftUI

// MARK: - investor lens

/// Value or Growth, with the core's own description of each. Used at onboarding and in Settings,
/// so the two can never describe the lens differently.
struct LensPicker: View {
    @Binding var selection: InvestorLens
    let lenses: [LensInfo]
    /// Onboarding shows each lens as a large card; Settings uses a compact segmented control.
    var style: Style = .compact
    var onChoose: ((InvestorLens) -> Void)? = nil

    enum Style { case compact, cards }

    var body: some View {
        switch style {
        case .compact: compact
        case .cards: cards
        }
    }

    private var compact: some View {
        VStack(alignment: .leading, spacing: 8) {
            Picker("Investor lens", selection: $selection) {
                ForEach(lenses) { Text($0.name).tag($0.lens) }
            }
            .pickerStyle(.segmented)
            if let current = lenses.first(where: { $0.lens == selection }) {
                Text(current.blurb)
                    .font(.caption)
                    .foregroundStyle(Theme.textSecondary)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
    }

    private var cards: some View {
        VStack(spacing: 12) {
            ForEach(lenses) { info in
                Button {
                    selection = info.lens
                    onChoose?(info.lens)
                } label: {
                    HStack(alignment: .center, spacing: 12) {
                        VStack(alignment: .leading, spacing: 6) {
                            Text(info.name).font(.vlHeadline).foregroundStyle(Theme.textPrimary)
                            Text(info.blurb).font(.subheadline).foregroundStyle(Theme.textSecondary)
                                .fixedSize(horizontal: false, vertical: true)
                        }
                        Spacer(minLength: 0)
                        Image(systemName: selection == info.lens ? "checkmark.circle.fill" : "chevron.right")
                            .foregroundStyle(selection == info.lens ? Theme.accent : Theme.textTertiary)
                    }
                    .padding(16)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .background(Theme.surface, in: RoundedRectangle(cornerRadius: 16, style: .continuous))
                    .overlay(RoundedRectangle(cornerRadius: 16, style: .continuous)
                        .stroke(selection == info.lens ? Theme.accent : Theme.border, lineWidth: selection == info.lens ? 2 : 1))
                }
                .buttonStyle(.plain)
                .accessibilityLabel("\(info.name). \(info.blurb)")
                .accessibilityAddTraits(selection == info.lens ? .isSelected : [])
            }
        }
    }
}

// MARK: - layout

/// The two layouts, each drawn by the layout itself on a bundled sample and scaled down — so what
/// you see here is what you will get, in your current theme and accent, not an illustration of it.
struct LayoutPicker: View {
    @Binding var selection: LayoutStyle
    let lens: InvestorLens

    @State private var sample: ValuationReport?
    @State private var explain: ExplainSummary?
    @State private var card: ReportCardSummary?

    var body: some View {
        HStack(alignment: .top, spacing: 14) {
            ForEach(LayoutStyle.allCases, id: \.self) { style in
                Button { selection = style } label: {
                    VStack(spacing: 8) {
                        thumbnail(style)
                            .overlay(RoundedRectangle(cornerRadius: 10, style: .continuous)
                                .stroke(selection == style ? Theme.accent : Theme.border, lineWidth: selection == style ? 2.5 : 1))
                        HStack(spacing: 4) {
                            if selection == style {
                                Image(systemName: "checkmark.circle.fill").foregroundStyle(Theme.accent)
                            }
                            Text(style.displayName).font(.subheadline.weight(.semibold)).foregroundStyle(Theme.textPrimary)
                        }
                        Text(style.blurb).font(.caption2).foregroundStyle(Theme.textSecondary)
                            .multilineTextAlignment(.center).fixedSize(horizontal: false, vertical: true)
                    }
                    .frame(maxWidth: .infinity)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("\(style.displayName) layout. \(style.blurb)")
                .accessibilityAddTraits(selection == style ? .isSelected : [])
            }
        }
        .padding(.vertical, 6)
        .task(id: lens) { await loadSample() }
    }

    /// Phone-width layout, cropped to its first screen, scaled to fit a thumbnail.
    private static let phoneWidth: CGFloat = 393
    private static let cropHeight: CGFloat = 640
    private static let scale: CGFloat = 0.34

    @ViewBuilder
    private func thumbnail(_ style: LayoutStyle) -> some View {
        let w = Self.phoneWidth * Self.scale, h = Self.cropHeight * Self.scale
        ZStack(alignment: .topLeading) {
            Theme.background
            if let sample {
                layout(style, sample)
                    .frame(width: Self.phoneWidth)
                    .frame(width: Self.phoneWidth, height: Self.cropHeight, alignment: .top)
                    .clipped()
                    .scaleEffect(Self.scale, anchor: .topLeading)
            } else {
                ProgressView().frame(width: w, height: h)
            }
        }
        .frame(width: w, height: h, alignment: .topLeading)
        .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
        .allowsHitTesting(false)
        .accessibilityHidden(true)   // the label below describes it
    }

    @ViewBuilder
    private func layout(_ style: LayoutStyle, _ report: ValuationReport) -> some View {
        let ctx = LayoutContext(
            report: report,
            result: report.result(for: .traditional),
            explain: explain,
            expert: false,
            showingMathTemporarily: false,
            model: .constant(.traditional),
            editPrice: {},
            setShowMath: { _ in },
            expandAll: .constant(nil),
            expandVersion: .constant(0),
            reportIssue: {},
            reportCard: card
        )
        switch style {
        case .classic: ClassicLayout(ctx: ctx)
        case .reportCard: ReportCardLayout(ctx: ctx)
        }
    }

    private func loadSample() async {
        // The bundled Apple sample: always present, needs no network or identity.
        let repo = SampleValuationRepository()
        guard let r = try? await repo.valuation(ticker: "AAPL", priceOverride: nil, overrides: .none) else { return }
        async let e = repo.explain(r)
        async let c = repo.reportCard(r, lens: lens)
        let (ex, rc) = await (e, c)
        sample = r; explain = ex; card = rc
    }
}
